tachyon/crypto/sumcheck/sumcheck_prover.h

```diff
@@ -0,0 +1,124 @@
+// Copyright 2022 arkworks contributors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.arkworks and the LICENCE-APACHE.arkworks
+// file.
+
+#ifndef TACHYON_CRYPTO_SUMCHECK_SUMCHECK_PROVER_H_
+#define TACHYON_CRYPTO_SUMCHECK_SUMCHECK_PROVER_H_
+
+#include <utility>
+#include <vector>
+
+#include "tachyon/crypto/sumcheck/sumcheck_prover_msg.h"
+#include "tachyon/crypto/sumcheck/sumcheck_proving_key.h"
+#include "tachyon/crypto/sumcheck/sumcheck_verifier_msg.h"
+#include "tachyon/math/polynomials/multivariate/linear_combination.h"
+
+namespace tachyon::crypto {
+
+// The prover argues for the sum of the polynomial over {0,1}^|num_variables_|
+//
+// The polynomial is represented by a list of terms each containing a
+// coefficient and a list of |MultilinearDenseEvaluations|. Refer to
+// "tachyon/math/polynomials/multivariate/linear_combination.h" for more info.
+template <typename MLE>
+class SumcheckProver {
+ public:
+  using F = typename MLE::Field;
+  using Term = math::LinearCombinationTerm<F>;
+  using Point = std::vector<F>;
+
+  SumcheckProver() = default;
+  explicit SumcheckProver(const SumcheckProvingKey<MLE>& key)
+      : num_variables_(key.verifying_key.num_variables),
+        max_evaluations_(key.verifying_key.max_evaluations),
+        terms_(key.terms),
+        flattened_ml_evaluations_(key.flattened_ml_evaluations) {
```

Comment:
Do you intend to copy here?

Reply:
from SumcheckProvingKey:
```
static SumcheckProvingKey Build(
      const math::LinearCombination<MLE>& linear_combination) {
    std::vector<MLE> flattened_ml_evaluations =
        base::Map(linear_combination.flattened_ml_evaluations(),
                  [](const std::shared_ptr<MLE> ptr) { return *ptr.get(); });
    return {SumcheckVerifyingKey::Build(linear_combination),
            linear_combination.terms(), std::move(flattened_ml_evaluations)};
  }
```
flattened_ml_evaluations is moved beforehand!

Reply:
Nevertheless, both `SumcheckProver` and `SumcheckProvingKey` have their own `terms` and `flattened_ml_evaluations` because the constructor copies them.